Paper: Ocetrac: An Object-Based Framework for Tracking and Quantifying Climate Events in Gridded Datasets

CassiaCai opened the pull request:
First draft of "Ocetrac: An Object-based Framework for Tracking and Quantifying Climate Events in Gridded Datasets"

CassiaCai committed:
initial changes

CassiaCai committed:
edits

CassiaCai committed:
added references in .bib, in-text citations...

CassiaCai committed:
references fix

CassiaCai committed:
table formatting

CassiaCai committed:
title

CassiaCai committed:
myst changes

CassiaCai committed:
discussion section

CassiaCai committed:
discussion section

CassiaCai committed:
banner

CassiaCai committed:
banner

CassiaCai committed:
banner again

CassiaCai committed:
banner without gridlines

CassiaCai committed:
section 2

CassiaCai committed:
banner

CassiaCai committed:
banner

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+# Ensure that this title is the same as the one in `myst.yml`
+title: 'Ocetrac: An Object-based Framework for Tracking and Quantifying Climate Events in Gridded Datasets'
+abstract: |
+  Climate extremes such as marine heatwaves and ocean acidification events appear as dynamic, irregular structures in gridded datasets. Software tools are available for detecting these events. However, there is a need for standardized methods for tracking their spatiotemporal evolution and evolving characteristics (such as intensity, shape, and duration). Ocetrac is an open-sourced Python package ...
```
when you introduce scikit learn and xarray first you can link to their reference.

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+## Introduction
+Gridded climate datasets are important for studying phenomena like marine heatwaves (MHWs), deoxygenation zones, and atmospheric blocking events among others. These events can exhibit complex spatiotemporal behavior including splitting, merging, and complex evolution. The ecological impact of these events are closely tied to their spatial and temporal characteristics. For example, in the case of  ...
+
+Conventional analysis approaches, including gridpoint-based statistics, fixed-region summaries, and empirical orthogonal functions (EOFs), offer valuable insight in explaining spatial patterns but fall short in characterizing events as coherent and evolving structures. Gridpoint statistical analysis obscures spatial connectivity while EOFs lack object-level granularity. Object tracking is a diffe ...
```
time is (t)

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+  Climate extremes such as marine heatwaves and ocean acidification events appear as dynamic, irregular structures in gridded datasets. Software tools are available for detecting these events. However, there is a need for standardized methods for tracking their spatiotemporal evolution and evolving characteristics (such as intensity, shape, and duration). Ocetrac is an open-sourced Python package ...
+---
+## Introduction
+Gridded climate datasets are important for studying phenomena like marine heatwaves (MHWs), deoxygenation zones, and atmospheric blocking events among others. These events can exhibit complex spatiotemporal behavior including splitting, merging, and complex evolution. The ecological impact of these events are closely tied to their spatial and temporal characteristics. For example, in the case of  ...
```
These impacts can include habitat destruction

valentina-s replied:
there is extra space before the last period.

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+
+Conventional analysis approaches, including gridpoint-based statistics, fixed-region summaries, and empirical orthogonal functions (EOFs), offer valuable insight in explaining spatial patterns but fall short in characterizing events as coherent and evolving structures. Gridpoint statistical analysis obscures spatial connectivity while EOFs lack object-level granularity. Object tracking is a diffe ...
+
+Here, we describe [Ocetrac](https://ocetrac.readthedocs.io/en/latest/), an open-source Python package designed for detecting, tracking, and quantifying irregular, evolving features in gridded climate data. Ocetrac combines morphological image processing, via `scikit-image` [@van2014scikit], with labelled array workflows, via `xarray` [@xarray], to characterize irregular structures. The Measures s ...
```
Here is the first time you are introducing the Measures submodule so it would be good to add more direct description (before you jump into the aggregation aspect). For example:

The Measures submodule provides functionality to quantify spatial and temporal properties of the tracked events, and thus facilitates further statistical analysis.

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+1. Detection: Ocetrac smooths spatially connected regions
+2. Tracking: Ocetrac labels objects across time steps, even when they split or merge
+2. Quantification: Ocetrac computes shape (e.g., compactness), motion (e.g., centroid velocity), and intensity (e.g., thermal anomaly) measures
+3. Integration: Ocetrac is compatible with the Scientific Python ecosystem, including `dask` [@dask] and  `scikit-learn` [@sklearn1]
```
add xarray

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+2. Quantification: Ocetrac computes shape (e.g., compactness), motion (e.g., centroid velocity), and intensity (e.g., thermal anomaly) measures
+3. Integration: Ocetrac is compatible with the Scientific Python ecosystem, including `dask` [@dask] and  `scikit-learn` [@sklearn1]
+
+Although originally developed for MHW analysis, Ocetrac can be used for other applications, from atmospheric rivers to hypoxic zones and bloom patches in the ocean. Its modular design allows users to customize detection thresholds, tracking logic, and meastures of tracked objects. This paper details its design, demonstrates several use cases, and outlines future directions. 
```
meastures -> measures

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+:::
+
+### Input Specifications and Preprocessing
+Ocetrac requires input data as a 3D `xarray.DataArray` with dimensions ordered as $(t, y, x)$, where $(y, x)$ correspond to regular spatial coordinates, typically latitude and longitude, and $t$ corresponds to time. The time dimension should be uniformly spaced to ensure optimal tracking performance of features through the temporal sequence. The time dimension should also have high enough spatial ...
```
There is something wrong with this sentence, since the time dimension does not have spatial dimension?
The time dimension should also have high enough spatial and temporal resolution to capture the temporal evolution of the target feature.

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+An optional binary land mask (1 for valid grid cells, 0 for excluded regions such as land or sea ice) can be provided to omit specific areas from detection and tracking. All preprocessing, such as detrending and anomaly calculation, thresholding, must be performed before using Ocetrac. Common thresholding approaches include: percentile-based (e.g., values exceeding the 90th percentile of anomalie ...
+
+### Tracking Submodule
+The process begins with converting anomalies into binary maps using a threshold (typically the 90th percentile for MHWs). Grid cells exceeding the threshold are marked as features where active grid cells are assigned the value of 1 and all others are inactive and assigned the value of 0.
```
The workflow begins with ...

valentina-s commented on papers/cassiacai/main.md:
````diff
@@ -0,0 +1,214 @@
+   lat_resolution_value=lat_resolution_value
+)
+```
+
````
Could show the structure of the output.

CassiaCai committed:
some changes

CassiaCai committed:
typo

valentina-s commented on papers/cassiacai/main.md:
````diff
@@ -0,0 +1,214 @@
+mo = obj_Tracker._morphological_operations()
+```
+Here Ocetrac provides basic diagnostics (e.g., initial/final object counts, area retention percentage) to evaluate tracking performance and feature characteristics. 
+
````
Could show the example structure of output.

freemansw1 replied:
I agree here. This is a major issue in the community right now.

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+Here Ocetrac provides basic diagnostics (e.g., initial/final object counts, area retention percentage) to evaluate tracking performance and feature characteristics. 
+
+### Measures Submodule
+The output of the tracker is a `xarray.Dataset` with many labelled events. The Measures submodule allows users to process large numbers of tracked events by automating repetitive calculations. Its standardized outputs allow for direct comparison across different events and datasets, which streamlines analysis workflows. The submodule supports both individual case studies and ensemble analyses. In ...
```
Move the sentence "the measures submodule processes the output of Ocetrac..." to the top and then discuss "large number of tracked events"

valentina-s replied:
shape, motion, intensity, and temporal, and contextual -> shape, motion, intensity, temporal, and contextual 

(remove one and)

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+  \endhead
+  
+  \hline
+  \multirow{5}{*}{SHAPE} & Area (Extent) & Spatial coverage in km² or grid cells. Calculated as the sum of cell areas within bounds. & Larger values = greater spatial influence \\
```
explain that is added across time.

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+  \hline
+  \multirow{5}{*}{SHAPE} & Area (Extent) & Spatial coverage in km² or grid cells. Calculated as the sum of cell areas within bounds. & Larger values = greater spatial influence \\
+  \cline{2-4}
+  & Perimeter & Boundary length (km), computed via geodesic contours. Calculated as Haversine sum of contour points. & Complex shapes yield high perimeters. \\
```
Calculated using the Haverine distance between the contour points. 

(Should we mention that we extract contours first?)

valentina-s commented on papers/cassiacai/main.md:
````diff
@@ -0,0 +1,214 @@
+```
+
+The computed measures are designed for application across both individual events and multiple events. Each measure is stored in a structured Python dictionary keyed by event ID, supporting single-event analysis and multi-event comparison, where ensemble statistics can be calculated when aggregating across multiple event IDs. The submodule includes built-in visualization tools that generate plots  ...
+
````
Show an example of calling one/two measure.

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+
+:::{figure} jtech_workflow.jpg
+:label: fig:JTECH
+Ocetrac workflow for MHW analysis in an observational data product (in this example, the [NOAA 1/4° Optimum Interpolation Sea Surface Temperature dataset](https://www.ncei.noaa.gov/products/optimum-interpolation-sst)). The diagram illustrates the processing pipeline from model output through event detection, tracking, and characterization of observational MHWs. 
```
Say something about how the measures were used: length of event?

valentina-s replied:
"the" before Mediterranean is repeated. Could remove the "Also."

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+
+:::{figure} lens_workflow.jpg
+:label: fig:LENS
+Ocetrac workflow for MHW analysis in climate ensembles. This approach allows for statistical analysis of event properties.
```
Point in the workflow the next stage: in this case the features go into the clustering algorithm.

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+:::
+
+### Comparing Observed Marine Heatwaves Against Many Simulated Events
+This example uses Ocetrac to compare MHWs in the observational record (i.e., the spatially large and long-lasting Northeast Pacific MHW ‘The Blob’ in 2014-2016) with events simulated in a large ensemble of climate models, such as the [Community Earth System Model version 2 Large Ensemble Community Project (CESM2-LENS)](https://www.cesm.ucar.edu/community-projects/lens2) [@Rod21]. The tool provide ...
```
This example uses ->
In this scenario, we show the use of Ocetrac to compare ...

valentina-s replied:
(Instead of tool) This approach facilitates a probabilistic framework ...

valentina-s replied:
Maybe statistical framework, instead of "probabilistic framework"

valentina-s commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,214 @@
+Object-based analysis of MHWs can be applied not only to observations and model output, but also to seasonal forecasts. In recent work, @Coh25 combine Ocetrac with the [Method for Object-based Diagnostic Evaluation (MODE)](https://metplus.readthedocs.io/projects/met/en/latest/Users_Guide/mode.html) to evaluate MHW forecasts. More information about MODE can be found in @Dav06a, @Dav06b, and @Bul16 ...
+
+## Discussion
+Ocetrac is an open-source software suite designed to identify, track, and quantify irregular, evolving features in gridded climate data, such as MHWs. It is actively being developed with extensibility in mind. Future plans include: (1) expanding grid compatibility (e.g., [Parallel Ocean Program (POP) grid](https://www.cesm.ucar.edu/models/pop) via helper functions, (2) enhancing tracking to incor ...
```
* remove "via helper functions"
* 2) enhancing tracking to incorporate vertical structure (remove "of")
* 4) streamlining workflow for improved usability: This goal might sound vague? Could remove it or add something more specific.

* remove "clearer documentation"

CassiaCai committed:
edits

CassiaCai committed:
edits

CassiaCai committed:
edit

CassiaCai committed:
edit

CassiaCai committed:
4.4

CassiaCai committed:
affil

CassiaCai committed:
affil

CassiaCai committed:
description

ameyxd commented:
Inviting reviewers: @shahchiragh and @shikoh8318@gmail.com

shahchiragh commented:
Hi @CassiaCai and team, I'm glad to be part of the review process for this paper. I’ll begin providing feedback over the next few days. Looking forward to engaging with your work and contributing to the review.

ameyxd commented:
@freemansw1 will serve as editor for this paper.

freemansw1 commented:
I missed that this manuscript needs two reviewers still, apologies. I've remarked it and am working on assigning new reviewers. @shahchiragh do you still plan to review?

shahchiragh commented:
Hi @freemansw1, apologies for the delay in posting my review comments. I'm sharing them here now for your reference. 

@CassiaCai -- overall, the paper meets most of the review criteria. Some minor comments and suggestions: 
- When introducing technical terms or acronyms (e.g., "CF-compliant NetCDF"), consider providing a brief explanation or reference to help readers who may not be familiar with these concepts (Section 2)
- Consider adding brief explanation of the rationale behind the choice of morphological operations used in Ocetrac (Section 3.2)
- Provide context for specific measures included in Measures submodule (Section 3.3)
- Some minor typos/errors:
   - should "habitat destructution" be "habitat destruction" (Section 1, "Introduction")?
   - "meastures" should be "measures" (Section 1, "Introduction")
   - there are instances where the narrative switches between different tenses or persons (Section 1,2, & 3)
  
While these issues are minor, addressing them can improve the paper's overall clarity, readability, and quality.

CassiaCai committed:
Reviewer comment edits

CassiaCai committed:
Removing figures PDF version

CassiaCai commented:
Hi @shahchiragh, thank you for your thoughtful feedback! We have addressed each of your comments.

**When introducing technical terms or acronyms (e.g., "CF-compliant NetCDF"), consider providing a brief explanation or reference to help readers who may not be familiar with these concepts (Section 2)**
We have added explanations for technical terms, like clarifying that "CF-compliant NCDF" refers to files following Climate and Forecast metadata conventions. We have also added more description of EOF analysis.

**Consider adding brief explanation of the rationale behind the choice of morphological operations used in Ocetrac (Section 3.2)**
We have expanded the discussion of morphological operations to explain why closing was chosen to maintain spatial coherence and why opening follows closing to remove artifacts. We have also added a note that other sequences can be used. Here is the begining of the changed paragraph: "The choice of these morphological operations follows image processing practices..."

**Provide context for specific measures included in Measures submodule (Section 3.3)**
We have added a rationale for each measure type in the paragraph that begins like this "Shape measures characterize geometric properties..."

**Minor typos/errors**
Typos (habitat destructution --> habitat destruction; meastures --> measures) have been fixed. Narrative switches between different tenses and persons have also been fixed.

CassiaCai committed:
changes to manuscript

CassiaCai committed:
typos

CassiaCai committed:
typos

CassiaCai committed:
typos

CassiaCai committed:
typos

CassiaCai committed:
typos

freemansw1 commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,240 @@
+Conventional analysis approaches, including gridpoint-based statistics, fixed-region summaries, and empirical orthogonal functions (EOFs), offer valuable insight in explaining spatial patterns but fall short in characterizing events as coherent and evolving structures. Gridpoint statistical analysis obscures spatial connectivity while EOFs, which is a dimensionality reduction technique that decom ...
+
+This paper describes [Ocetrac](https://github.com/ocetrac/ocetrac/releases/tag/v0.1.6), an open-source Python package designed for detecting, tracking, and quantifying irregular, evolving features in gridded climate data. Ocetrac combines morphological image processing using `scikit-image` [@van2014scikit] with labelled array workflows using `xarray` [@xarray] to characterize irregular structures ...
+1. Detection: Ocetrac smooths spatially connected regions (features) in a binary field, such as sea surface temperature (SST) anomalies, to identify objects that exceed a user-defined threshold (e.g., the 90th percentile of anomalies)
```
I think the way that this is worded can leave the reader a bit confused. In the previous paragraph, you note that TempestExtremes (and indeed most tracking algorithms) uses a threshold-based approach, which has limitations, and here you are describing an approach that is also threshold-based. I understand that Ocetrack avoids some of the normal pitfalls of typical threshold-based metrics, so I'd just add in some language here that links that in.

CassiaCai replied:
Thank you for bringing this up. We have added "Although Ocetrac uses an initial threshold to identify candidate regions, Ocetrac moves beyond a simple binary definition by using morphological operations to filter and refine these regions based on their spatial coherence and connectivity."

freemansw1 commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,240 @@
+This paper describes [Ocetrac](https://github.com/ocetrac/ocetrac/releases/tag/v0.1.6), an open-source Python package designed for detecting, tracking, and quantifying irregular, evolving features in gridded climate data. Ocetrac combines morphological image processing using `scikit-image` [@van2014scikit] with labelled array workflows using `xarray` [@xarray] to characterize irregular structures ...
+1. Detection: Ocetrac smooths spatially connected regions (features) in a binary field, such as sea surface temperature (SST) anomalies, to identify objects that exceed a user-defined threshold (e.g., the 90th percentile of anomalies)
+2. Tracking: Ocetrac labels objects across time steps, even when they split or merge
+2. Quantification: Ocetrac computes shape, motion, and intensity measures
```
2 twice in a row

CassiaCai replied:
The text has been revised to show 2 only once.

freemansw1 commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,240 @@
+Although originally developed for MHW analysis, Ocetrac can be used for other applications, from atmospheric rivers to hypoxic zones and bloom patches in the ocean. Its modular design allows users to customize detection thresholds, tracking logic, and measures of tracked objects. This paper details its design, demonstrates several use cases, and outlines future directions. 
+
+## Goals and Motivations
+Climate phenomena like MHWs can exhibit dynamic spatial coherence as they evolve across space and time. Traditional gridpoint or static regional analyses that evaluate variables at individual locations or within a fixed region may overlook the structure and the evolution of such events. Object-based methods address this limitation by treating features as one event. This approach shifts the focus  ...
```
"dynamic spatial coherence" is jargon-y, suggest changing.

CassiaCai replied:
Thank you for this suggestion. This has been replaced with: "Climate phenomena like MHWs are dynamic events that can propagate across basins, changing their size, shape, and intensity as they evolve over time."

freemansw1 commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,240 @@
+## Goals and Motivations
+Climate phenomena like MHWs can exhibit dynamic spatial coherence as they evolve across space and time. Traditional gridpoint or static regional analyses that evaluate variables at individual locations or within a fixed region may overlook the structure and the evolution of such events. Object-based methods address this limitation by treating features as one event. This approach shifts the focus  ...
+
+Object-based detection and tracking of spatiotemporal climate features, specifically MHWs, is a growing area of research. Other climate features, such as tropical cyclones and atmospheric rivers, have benefitted from tracking algorithms [@Ull21]. Recent studies have used a variety of techniques to identify MHW events, each with its own benefits and limitations. All techniques track objects using  ...
```
I think some of this should move back to the introduction. You also might take a look at the toocan algorithm, which (although the code is not open iirc) sounds somewhat similar here (although targeted at atmospheric science): https://essd.copernicus.org/articles/16/4021/2024/

CassiaCai replied:
Thank you for this suggestion. I have added a sentence about TOOCAN. This section has been moved to the Introduction. 2 paragraphs that were in "Goals and Motivations" have been revised and moved to "Introduction."

freemansw1 commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,240 @@
+
+The choice of these morphological operations follows image processing practices for feature enhancement and noise reduction. The closing operation is applied first to connect nearby features, followed by opening to remove small artifacts and refine the boundaries of the features. This sequence is important for maintaining the integrity of the detected features while minimizing fragmentation. The  ...
+
+The operations use a circular structuring element with radius $R$ (in grid cells). For 0.25° resolution data:
```
Define what a circular structuring element is here.

freemansw1 replied:
Also, please note what the resolution dependence is here. If someone had 0.125° data, would it scale by a factor of 2?

CassiaCai replied:
We have now included a description of the circular structuring element.  We have also: "For data at a different resolution, the radius must be scaled to maintain a consistent physical filtering scale. For example, for higher-resolution 0.125° data, the grid cell size is halved. To achieve an equivalent physical filter size, $R$ should be doubled. "

freemansw1 commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,240 @@
+
+The choice of $R$ represents a trade-off between feature completeness (higher $R$ retains more connected areas) and spatial precision (lower $R$ preserves finer-scale details) ([Figure %s](#fig:radius_size)). Users should validate this parameter against known feature scales in their domain. Features, after undergoing morphological operations, are now referred to as objects.
+
+Ocetrac then applies size-based filtering to focus on the most spatially coherent objects. It calculates the area of each detected object in grid cells, then removes objects smaller than a user-defined area threshold $P$. By default, it keeps objects larger than 75% of all detected features (the 75th percentile), meaning the smallest 25% of objects are filtered out. This filtering step helps elim ...
```
Is the percentile here a per-time percentile, or an overall percentile for the full dataset?

CassiaCai replied:
We have clarified using "for the full dataset"

freemansw1 commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,240 @@
+
+:::{figure} radsize.jpg
+:label: fig:radius_size
+Objects outlined (purple outlines) detected using varying $R$ values ($R$ ranging from 1° to 7°). The background shows SST anomalies (°C) from February 2007 of an ensemble member of the CESM2-LENS dataset.
```
Note the grid spacing here

CassiaCai replied:
Added "The horizontal resolution is 1°."

freemansw1 commented on papers/cassiacai/main.md:
```diff
@@ -0,0 +1,240 @@
+Objects outlined (purple outlines) detected using varying $R$ values ($R$ ranging from 1° to 7°). The background shows SST anomalies (°C) from February 2007 of an ensemble member of the CESM2-LENS dataset.
+:::
+
+Ocetrac identifies objects in each time step by grouping together adjacent pixels, including diagonal connections (8-connectivity). Then, it matches objects that appear in either the exact same grid cell or directly adjacent cells (up/down/left/right) in the next timestep (6-connectivity). This conservative approach allows objects to move gradually and prevents discontinuous jumps. The system mai ...
```
I think an increased description of why 8-connectivity for feature detection was used vs the 6-connectivity for tracking.

freemansw1 replied:
Does Ocetrack work across the poles?

CassiaCai replied:
We have added a more detailed description of why 8-connectivity is first used for feature detection followed by 6-connectivity for the temporal tracking. The current implementation of Ocetrac is designed for rectilinear grids and successfully tracks features within the Arctic and Antarctic regions. We have also added a note about tracking at high latitudes: "The current implementation of Ocetrac is designed for rectilinear (latitude-longitude) grids. A limitation of this framework is its handling of high-latitude regions. At high latitudes, the convergence of meridians leads to distortions in area representation, which can affect the accuracy of object detection and tracking. Future versions of Ocetrac will address this limitation by incorporating grid-aware algorithms that adjust for varying grid cell sizes and shapes at different latitudes."

freemansw1 reviewed (commented):
For the sake of expediency, I will serve as the second reviewer.

Overall, the manuscript is in fantastic shape. Very well done!! Ocetrack is an important addition to the earth science tracking community, and this is an excellent introduction to it. 

I have some minor comments in the manuscript. The biggest one for me is a description of the output format- handling lagrangian output in usable, reproducible data structures is a key issue in the community right now.

shahchiragh reviewed (commented):
@CassiaCai thank you for effectively addressing all the concerns and suggestions provided in the previous review. The changes made by the authors have improved the clarity, readability, and overall quality of the paper, and it is now considered complete from my end.

CassiaCai committed:
reviewer 2 edits

CassiaCai committed:
reviewer 2 edits

CassiaCai committed:
reviewer 2 edits

freemansw1 commented:
@CassiaCai Thanks much for the great edits. I'm happy with this after your revisions. Nice work!
